.github/scripts/trymerge.py

```diff
@@ -1420,24 +1486,33 @@ def merge(pr_num: int, repo: GitRepo,
                 pr.delete_land_time_check_branch(repo)
             raise RuntimeError("New commits were pushed while merging. Please rerun the merge command.")
         try:
-            find_matching_merge_rule(pr, repo)
-            checks = get_combined_checks_from_pr_and_land_validation(pr, land_check_commit)
-            pending = filter_pending_checks(checks)
-            failing = filter_failed_checks(checks)
+            required_checks = []
+            failed_rule_message = None
+            try:
+                find_matching_merge_rule(pr, repo)
```

Comment:
If I read the code correctly, this would raise `MandatoryChecksMissingError` when there are pending jobs.  So how are the 2 branches of `failed_rule_message is not None` run on line 1511?

I guess the question is about the exceptions raised by `find_matching_merge_rule` (also calls `get_combined_checks_from_pr_and_land_validation` inside) v.s. the exceptions raising here in `merge`.  They looks duplicated.  But then trymerge logic is complicated, so I wonder what I miss here

Reply:
not entirely sure if i understand the first question, but some None checks are just in case.

  For the exception raising duplication, let's say a pull job is pending, but a trunk job has failed.  Then `find_matching_merge_rule` doesn't see the failing trunk job because it only checks the mandatory jobs (none of which include trunk at the moment).  Then it raises `MandatoryChecksMissingError` which previously would get caught at the end of the loop, and we would wait 5 minutes and run the loop again.  By catching the `MandatoryChecksMissingError` earlier, we can also check the non mandatory checks (trunk) and raise RuntimeError, which would cause the trymerge script to exit early.

`get_combined_checks_from_pr_and_land_validation` does get called twice but `get_checkrun_conclusions` does check for self.conclusions first, so at least its not rerunning all the code.

Reply:
Got it, so `find_matching_merge_rule` is for mandatory checks while `merge` will handle both mandatory and non-mandatory ones.  The latter behavior is what I'm familiar with, so I wrongly thought that `find_matching_merge_rule` did both.  This also answers my first question.

Reply:
Per discussion with @clee2000, may be we should just get rid of the obsolete distinction between mandatory and non-mandatory checks in trymerge.  So the logic could be as simple as:

* Are pull and trunk run?
  * Yes, continue
  * No, the merge process stops and clearly says the reason
* No job fails.  All is good and the commit will be merged
* Some jobs fail.  Are they ignorable?
   * Yes, the commit will be merged
   * No, the merge process stops and clearly says the reason